File: Medium/1024 - VideoStitching/solution.ts
// Question 1024. Video Stitching (https://leetcode-cn.com/problems/video-stitching/)

/**
 * This solution will failed in the case where max end time of the clip is greater than
 * the target value T, but the clips are not overlapped. e.g.
 *  [[0,2],[4,8]]
 *  5
 *
 * @param {number[][]} clips
 * @param {number} T
 * @returns {number}
 */
function videoStitchingFailed(clips: number[][], T: number): number {
  const clipsStarEndTimeMapping: Map<number, number[]> = new Map<
    number,
    number[]
  >();

  const startTimes: number[] = [];
  let maxEndTimes: number = 0;
  for (let [startTime, endTime] of clips) {
    if (clipsStarEndTimeMapping.has(startTime)) {
      const endTimeForTheStartTime: number[] = clipsStarEndTimeMapping.get(
        startTime
      ) as number[];
      clipsStarEndTimeMapping.set(startTime, [
        ...endTimeForTheStartTime,
        endTime,
      ]);
    } else {
      clipsStarEndTimeMapping.set(startTime, [endTime]);
      startTimes.push(startTime);
    }

    maxEndTimes = Math.max(maxEndTimes, endTime);
  }

  let startTime: number = 0;
  if (maxEndTimes < T || !clipsStarEndTimeMapping.has(startTime)) {
    return -1;
  }

  let clipsNumber: number = 1;
  let endTime: number = Math.max(
    ...(clipsStarEndTimeMapping.get(startTime) as number[])
  );

  while (
    endTime < T &&
    startTimes.filter((nextStartTime) => nextStartTime <= endTime)
  ) {
    let possibleStartTimes: number[] = startTimes.filter(
      (nextStartTime) => nextStartTime <= endTime
    ) as number[];

    // console.log(
    //   `endTime = ${endTime}, clipsNumber = ${clipsNumber}, startTimes = [${startTimes.join(
    //     ", "
    //   )}], possibleStartTimes = [${possibleStartTimes.join(", ")}]`
    // );

    endTime = Math.max(
      ...possibleStartTimes
        .map(
          (possibleStartTime) =>
            clipsStarEndTimeMapping.get(possibleStartTime) as number[]
        )
        .reduce((acc, current) => [...acc, ...current])
    );

    clipsNumber++;
  }

  return endTime >= T ? clipsNumber : -1;
}

/**
 * 执行用时：84 ms, 在所有 typescript 提交中击败了 100.00% 的用户
 * 内存消耗：39.7 MB, 在所有 typescript 提交中击败了 100.00% 的用户
 *
 * Acknowledgement: This solution borrows the lastEndTime idea from the 
 * [official solution](https://leetcode-cn.com/problems/video-stitching/solution/shi-pin-pin-jie-by-leetcode-solution/),
 * so the dilemma in teh function videoStitchingFailed can be resolved.
 *
 * @param {number[][]} clips
 * @param {number} T
 * @returns {number}
 */
function videoStitching(clips: number[][], T: number): number {
  const clipsStarEndTimeMapping: Map<number, number[]> = new Map<
    number,
    number[]
  >();

  const startTimes: number[] = [];
  let maxEndTimes: number = 0;
  for (let [startTime, endTime] of clips) {
    if (clipsStarEndTimeMapping.has(startTime)) {
      const endTimeForTheStartTime: number[] = clipsStarEndTimeMapping.get(
        startTime
      ) as number[];
      clipsStarEndTimeMapping.set(startTime, [
        ...endTimeForTheStartTime,
        endTime,
      ]);
    } else {
      clipsStarEndTimeMapping.set(startTime, [endTime]);
      startTimes.push(startTime);
    }

    maxEndTimes = Math.max(maxEndTimes, endTime);
  }

  let startTime: number = 0;
  if (maxEndTimes < T || !clipsStarEndTimeMapping.has(startTime)) {
    return -1;
  }

  let clipsNumber: number = 1;
  let endTime: number = Math.max(
    ...(clipsStarEndTimeMapping.get(startTime) as number[])
  );
  let lastEndTime: number = endTime;

  while (
    endTime < T &&
    startTimes.filter((nextStartTime) => nextStartTime <= endTime)
  ) {
    let possibleStartTimes: number[] = startTimes.filter(
      (nextStartTime) => nextStartTime <= endTime
    ) as number[];

    // console.log(
    //   `endTime = ${endTime}, clipsNumber = ${clipsNumber}, startTimes = [${startTimes.join(
    //     ", "
    //   )}], possibleStartTimes = [${possibleStartTimes.join(", ")}]`
    // );

    endTime = Math.max(
      ...possibleStartTimes
        .map(
          (possibleStartTime) =>
            clipsStarEndTimeMapping.get(possibleStartTime) as number[]
        )
        .reduce((acc, current) => [...acc, ...current])
    );

    if (lastEndTime === endTime) {
      return -1;
    }

    clipsNumber++;
  }

  return endTime >= T ? clipsNumber : -1;
}

File: Medium/1024 - VideoStitching/Solution.cs
// Question 1024. Video Stitching (https://leetcode-cn.com/problems/video-stitching/)

using System;

public class Solution
{
    /// <summary>
    /// 执行用时：100 ms, 在所有 C# 提交中击败了 100.00% 的用户
    /// 内存消耗：24.2 MB, 在所有 C# 提交中击败了 50.00% 的用户
    /// 
    /// Acknowledgement: This solution is just a C# version of the 
    /// [official solution](https://leetcode-cn.com/problems/video-stitching/solution/shi-pin-pin-jie-by-leetcode-solution/) 
    /// method II.
    /// </summary>
    /// <param name="clips"></param>
    /// <param name="T"></param>
    /// <returns></returns>
    public int VideoStitching(int[][] clips, int T)
    {
        int[] maxN = new int[T];
        foreach (int[] clip in clips)
        {
            int startTime = clip[0];
            int endTime = clip[1];

            if (startTime < T)
            {
                maxN[startTime] = Math.Max(maxN[startTime], endTime);
            }
        }

        int last = 0;
        int previousEnd = 0;
        int clipsCount = 0;

        for (int i = 0; i < T; i++)
        {
            last = Math.Max(last, maxN[i]);
            if (last == i)
            {
                return -1;
            }

            if (i == previousEnd)
            {
                clipsCount++;
                previousEnd = last;
            }
        }

        return clipsCount;
    }

    /// <summary>
    /// 执行用时：104 ms, 在所有 C# 提交中击败了 100.00% 的用户
    /// 内存消耗：24.5 MB, 在所有 C# 提交中击败了 25.00% 的用户
    /// 
    /// Acknowledgement: This solution is just a C# version of the 
    /// [official solution](https://leetcode-cn.com/problems/video-stitching/solution/shi-pin-pin-jie-by-leetcode-solution/) 
    /// method I.
    /// </summary>
    /// <param name="clips"></param>
    /// <param name="T"></param>
    /// <returns></returns>
    public int VideoStitchingDP(int[][] clips, int T)
    {
        int[] dp = new int[T + 1];
        Array.Fill(dp, int.MaxValue - 1);

        dp[0] = 0;

        for (int i = 1; i <= T; i++)
        {
            foreach (int[] clip in clips)
            {
                var startTime = clip[0];
                var endTime = clip[1];

                if (startTime < i && i <= endTime)
                {
                    dp[i] = Math.Min(dp[i], dp[startTime] + 1);
                }
            }
        }

        return dp[T] == int.MaxValue - 1 ? -1 : dp[T];
    }
}